test-framework/sudo-test/src/docker.rs

```diff
@@ -1,58 +1,51 @@
 use core::str;
 use std::{
-    fs,
-    process::{Command, ExitStatus, Stdio},
+    fs::{self, File},
+    io::{Seek, SeekFrom, Write},
+    path::PathBuf,
+    process::{Command as StdCommand, Stdio},
 };
 
 use tempfile::NamedTempFile;
 
-use crate::{helpers, Result};
+use crate::{Result, SudoUnderTest, BASE_IMAGE};
 
-const DEFAULT_COMMAND: &[&str] = &["sleep", "infinity"];
+pub use self::command::{Command, Output};
+
+mod command;
+
+const DOCKER_RUN_COMMAND: &[&str] = &["sleep", "infinity"];
 
 pub struct Container {
     id: String,
 }
 
 impl Container {
     pub fn new(image: &str) -> Result<Self> {
-        let mut cmd = Command::new("docker");
-        cmd.args(["run", "-d", "--rm", image]).args(DEFAULT_COMMAND);
-        let id = helpers::stdout(&mut cmd, None)?;
-        validate_docker_id(&id, &cmd)?;
+        let mut docker_run = StdCommand::new("docker");
+        docker_run
+            .args(["run", "-d", "--rm", image])
+            .args(DOCKER_RUN_COMMAND);
+        let id = run(&mut docker_run, None)?.stdout()?;
+        validate_docker_id(&id, &docker_run)?;
 
         Ok(Container { id })
     }
 
-    pub fn exec(
-        &self,
-        cmd: &[impl AsRef<str>],
-        user: As,
-        stdin: Option<&str>,
-    ) -> Result<ExecOutput> {
-        helpers::run(&mut self.docker_cmd(cmd, user, stdin.is_some()), stdin)
-    }
-
-    /// Returns `$cmd`'s stdout if it successfully exists
-    pub fn stdout(&self, cmd: &[impl AsRef<str>], user: As, stdin: Option<&str>) -> Result<String> {
-        helpers::stdout(&mut self.docker_cmd(cmd, user, stdin.is_some()), stdin)
-    }
-
-    fn docker_cmd(&self, cmd: &[impl AsRef<str>], user: As, with_stdin: bool) -> Command {
-        let mut docker_cmd = Command::new("docker");
-        docker_cmd.arg("exec");
-        if with_stdin {
-            docker_cmd.arg("-i");
-        }
-        if let Some(user) = user.as_string() {
-            docker_cmd.arg("--user");
-            docker_cmd.arg(user);
+    pub fn exec(&self, cmd: &Command) -> Result<Output> {
+        let mut docker_exec = StdCommand::new("docker");
+        docker_exec.arg("exec");
+        if cmd.get_stdin().is_some() {
+            docker_exec.arg("-i");
         }
-        docker_cmd.arg(&self.id);
-        for arg in cmd {
-            docker_cmd.arg(arg.as_ref());
+        if let Some(user) = cmd.get_user() {
+            docker_exec.arg("--user");
```

Comment:
This looks like a common operation. Maybe `Command` could have a method that receives the `StdCommand` and sets the user. What do you think?

Reply:
I want this PR merged so that can be a future refactor :-)